In [1]:
import pandas as pd

In [14]:
## read data function
def read_data():
    data=pd.read_csv(r"C:\Users\pv11379\OneDrive - Deere & Co\Desktop\Personal folder\Data science\Database/iphone_purchase_records.csv")
    return data

df=read_data()

In [92]:
## importing necessary liabraries
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler,MinMaxScaler,RobustScaler
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score,precision_score,confusion_matrix,classification_report,r2_score
from sklearn.ensemble import RandomForestClassifier
import time
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from math import sqrt
from sklearn.linear_model import LinearRegression

In [21]:
## convert categorical data into numeric data
gender={'Male':0,'Female':1}
df['Genders']=df['Gender'].map(gender)

In [68]:
## Random_forest classifier
def data_processing(df):
    x=df[['Genders','Age','Salary']]
    y=df['Purchase Iphone']
    sc=StandardScaler()
    sc_x=sc.fit_transform(x)
    x_train,x_test,y_train,y_test=train_test_split(sc_x,y,train_size=0.8,random_state=2020,stratify=y)
    return x_train,x_test,y_train,y_test

x_train,x_test,y_train,y_test=data_processing(df)

def model(x_train,x_test,y_train):
    model=RandomForestClassifier(criterion='entropy',max_depth=10)
    model.fit(x_train,y_train)
    pred=model.predict(x_test)
    return pred

pred=model(x_train,x_test,y_train)

def validation(y_test,pred):
    acc=accuracy_score(y_test,pred)
    conf=confusion_matrix(y_test,pred)
    pre=precision_score(y_test,pred)
    return acc,conf,pre

acc,conf,pre=validation(y_test,pred)

print("accuracy score is - ",acc)
print("precision score is- ",pre)
print("confusion metrix is- ",conf)

accuracy score is -  0.925
precision score is-  0.8484848484848485
confusion metrix is-  [[46  5]
 [ 1 28]]


In [71]:
## logistic regression 
def lg_model(x_train,x_test,y_train):
    model1=LogisticRegression()
    model1.fit(x_train,y_train)
    pred1=model1.predict(x_test)
    return pred1

pred1=lg_model(x_train,x_test,y_train)

def validation(y_test,pred1):
    acc1=accuracy_score(y_test,pred1)
    conf1=confusion_matrix(y_test,pred1)
    pre1=precision_score(y_test,pred1)
    return acc1,conf1,pre1

acc1,conf1,pre1=validation(y_test,pred1)

print("accuracy score is - ",acc1)
print("precision score is- ",pre1)
print("confusion metrix is- ",conf1)

accuracy score is -  0.8375
precision score is-  0.8333333333333334
confusion metrix is-  [[47  4]
 [ 9 20]]


In [88]:
## KNN Classifier
def k_value(df):
    k=sqrt(df.shape[0])
    return k
k=k_value(df)
print(k)

def knn_model(x_train,x_test,y_train):
    model2=KNeighborsClassifier(n_neighbors=21)
    model2.fit(x_train,y_train)
    pred2=model2.predict(x_test)
    return pred2

pred2=knn_model(x_train,x_test,y_train)

def validation(y_test,pred2):
    acc2=accuracy_score(y_test,pred2)
    conf2=confusion_matrix(y_test,pred2)
    pre2=precision_score(y_test,pred2)
    return acc2,conf2,pre2

acc2,conf2,pre2=validation(y_test,pred2)

print("accuracy score is - ",acc2)
print("precision score is- ",pre2)
print("confusion metrix is- ",conf2)

20.0
accuracy score is -  0.925
precision score is-  0.8709677419354839
confusion metrix is-  [[47  4]
 [ 2 27]]


In [96]:
## Multiple linear regression
def linear_model(x_train,x_test,y_train):
    model3=LinearRegression()
    model3.fit(x_train,y_train)
    pred3=model3.predict(x_test)
    print("intercept is-",model3.intercept_)
    print("coef is - ",model3.coef_)
    return pred3

pred3=linear_model(x_train,x_test,y_train)

def validation(y_test,pred3):
    score=r2_score(y_test,pred3)
    return score

score=validation(y_test,pred3)

print("score is - ",score)

intercept is- 0.3636584040185817
coef is -  [-0.0151891   0.27884494  0.13445458]
score is -  0.39078187748003856


In [102]:
## Descision tree classifier
def tr_model(x_train,x_test,y_train):
    model4=DecisionTreeClassifier(criterion='entropy',max_depth=5)
    model4.fit(x_train,y_train)
    pred4=model4.predict(x_test)
    return pred4

pred4=tr_model(x_train,x_test,y_train)

def validation(y_test,pred4):
    acc4=accuracy_score(y_test,pred4)
    conf4=confusion_matrix(y_test,pred4)
    pre4=precision_score(y_test,pred4)
    return acc4,conf4,pre4

acc4,conf4,pre4=validation(y_test,pred4)

print("accuracy score is - ",acc4)
print("precision score is- ",pre4)
print("confusion metrix is- ",conf4)

accuracy score is -  0.9125
precision score is-  0.84375
confusion metrix is-  [[46  5]
 [ 2 27]]


## Random forest classifier model having highest accuracy value i.e 92.5%. Hence this model is suitable for better prediction.